react-app/src/hooks/useSearchParamsObject.ts

```diff
@@ -0,0 +1,55 @@
+import { useCallback, useEffect, useMemo } from "react";
+import qs from "qs";
+import { useLocation, useSearchParams } from "react-router-dom";
+
+/**
+ * A enhanced useSearchParams that allows objects
+ * rather than only as a key string-value only store
+ *
+ * @param defaultParams - params to set when search string is empty
+ */
+const useSearchParamsObject = <T>(
+  defaultParams?: T
+): [searchParams: T, setSearchParams: (v: T | ((prev: T) => T)) => void] => {
+  const location = useLocation();
+  const [_, setSearchParamsString] = useSearchParams();
+
+  const searchParams: T = useMemo(() => {
+    const searchString = location.search.replace("?", "");
+    const params = qs.parse(searchString, {
+      decoder: (str, defaultDecoder, charset, type) => {
+        if (type === "value") {
+          if (str === "true" || str === "false") {
+            return Boolean(str);
+          }
+          if (!isNaN(parseInt(str, 10))) {
+            return parseInt(str, 10);
+          }
+        }
+        return defaultDecoder(str, null, charset);
+      },
+    });
+
+    return params as unknown as T;
+  }, [location.search]);
+
+  const setSearchParams = useCallback(
+    (val: T | ((prev: T) => T)) => {
+      const params = {
+        ...(val instanceof Function ? val(searchParams) : val),
+      };
+      setSearchParamsString(qs.stringify(params, { encode: false }));
+    },
+    [searchParams, setSearchParamsString]
+  );
+
+  useEffect(() => {
+    if (location.search.length <= 1 && defaultParams) {
+      setSearchParamsString(qs.stringify(defaultParams, { encode: false }));
+    }
```

Comment:
can use `setSearchParams`?

Reply:
As mentioned in https://github.com/oursky/likedao/pull/181#discussion_r905836426 useSearchParams is not enough for our use case (ie. having nested objects in search params), I defined a custom setSearchParams above. 

I think we can use `navigate` from `useNavigate()`, but probably not `setSearchParams` from `useSearchParams()`

Reply:
yes i am referring to the `setSearchParams` you defined above
it's just because 
```
setSearchParamsString(qs.stringify(params, { encode: false }));
```
and
```
setSearchParamsString(qs.stringify(defaultParams, { encode: false }));
```
look very similar